File: src/main/services/mod-tools/touch-profile-detection.ts
import path from "node:path";

import { loadIniBundle } from "@main/lib/mod-static-glb/ini-loader";
import fse from "fs-extra";

import {
    TOUCH_PROFILE_MANIFEST_FILE,
    TOUCH_PROFILE_MANIFEST_KIND,
    TOUCH_SHADER_FILES,
} from "./touch-profile-types";

export type TouchProfileInputStatus = "none" | "generated" | "incomplete" | "suspected";

export type TouchProfileInputDetection = {
    status: TouchProfileInputStatus;
    reasons: string[];
    namespace?: string;
};

export async function inspectTouchProfileInput(modPath: string) {
    const bundle = await loadIniBundle(modPath);
    return inspectTouchProfileBundle(path.dirname(bundle.iniPath), bundle.sourcePaths);
}

export async function inspectTouchProfileBundle(
    modRoot: string,
    sourcePaths: string[],
): Promise<TouchProfileInputDetection> {
    const root = path.resolve(modRoot);
    const manifest = await readManifest(path.join(root, TOUCH_PROFILE_MANIFEST_FILE));
    if (manifest) {
        return withRuntimeFileStatus(root, {
            status: "generated",
            reasons: ["Nahida Touch Profile manifest found"],
        });
    }

    const iniText = (
        await Promise.all(
            [...new Set(sourcePaths)].map((sourcePath) => fse.readFile(sourcePath, "utf8")),
        )
    ).join("\n");
    const stateNamespaces = extractNamespaces(
        iniText,
        /Nahida Touch Profile state\s*\(\s*([^\s)]+)\s*\)/gi,
    );
    const runtimeNamespaces = extractNamespaces(
        iniText,
        /Nahida Touch Profile runtime\s*\(\s*([^\s)]+)\s*\)/gi,
    );
    const matchingNamespace = stateNamespaces.find((namespace) =>
        runtimeNamespaces.includes(namespace),
    );

    if (matchingNamespace) {
        return withRuntimeFileStatus(root, {
            status: "generated",
            namespace: matchingNamespace,
            reasons: ["Nahida Touch Profile INI markers found"],
        });
    }

    if (stateNamespaces.length > 0 || runtimeNamespaces.length > 0) {
        return withRuntimeFileStatus(root, {
            status: "incomplete",
            reasons: ["Incomplete Nahida Touch Profile INI markers found"],
        });
    }

    const missingShaders = await missingRuntimeShaders(root);
    const hasRuntimeReferences = [
        "rzm_gs_probe.hlsl",
        "rzm_object_detect.hlsl",
        "rzm_jiggle_interaction.hlsl",
    ].every((fileName) => iniText.toLowerCase().includes(fileName.toLowerCase()));
    if (missingShaders.length === 0 && hasRuntimeReferences) {
        return {
            status: "suspected",
            reasons: ["Touch runtime shaders and INI references found"],
        };
    }

    return { status: "none", reasons: [] };
}

export async function assertTouchProfileInputAllowed(modPath: string) {
    const detection = await inspectTouchProfileInput(modPath);
    assertTouchProfileDetectionAllowed(modPath, detection);
}

export function assertTouchProfileDetectionAllowed(
    modPath: string,
    detection: TouchProfileInputDetection,
) {
    if (detection.status === "none") return;

    const code =
        detection.status === "suspected"
            ? "TOUCH_PROFILE_INPUT_SUSPECTED_TOUCH"
            : "TOUCH_PROFILE_INPUT_ALREADY_TOUCH";
    const message =
        detection.status === "suspected"
            ? "Input appears to be an existing Touch mod and cannot be safely converted."
            : "Input is already a Nahida Touch Profile mod and cannot be converted again.";
    const details = detection.reasons.length > 0 ? ` ${detection.reasons.join("; ")}.` : "";
    throw new Error(`${code}: ${message}${details} Path: ${path.resolve(modPath)}`);
}

async function readManifest(manifestPath: string) {
    if (!(await fse.pathExists(manifestPath))) return false;

    try {
        const manifest = await fse.readJson(manifestPath);
        return manifest?.kind === TOUCH_PROFILE_MANIFEST_KIND;
    } catch {
        return false;
    }
}

async function withRuntimeFileStatus(
    root: string,
    detection: TouchProfileInputDetection,
): Promise<TouchProfileInputDetection> {
    const missingShaders = await missingRuntimeShaders(root);
    if (missingShaders.length === 0) return detection;

    return {
        ...detection,
        status: "incomplete",
        reasons: [...detection.reasons, `Missing runtime shaders: ${missingShaders.join(", ")}`],
    };
}

async function missingRuntimeShaders(root: string) {
    const runtimeRoot = path.join(root, "Resources", "IM");
    const existing = await Promise.all(
        TOUCH_SHADER_FILES.map(async (fileName) =>
            (await fse.pathExists(path.join(runtimeRoot, fileName))) ? null : fileName,
        ),
    );
    return existing.filter((fileName) => fileName !== null);
}

function extractNamespaces(text: string, pattern: RegExp) {
    return [...text.matchAll(pattern)].map((match) => match[1].toLowerCase());
}

File: internal/tools/touch_profile_detection.go
package tools

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type touchInputDetection struct {
	Status    string
	Reasons   []string
	Namespace string
}

func assertTouchProfileInputAllowed(modPath string) error {
	_, _, sourcePaths, err := loadModINIBundleWithSources(modPath)
	if err != nil {
		return err
	}
	iniPath, err := findPrimaryModINI(modPath)
	if err != nil {
		return err
	}
	return assertTouchProfileBundleAllowed(filepath.Dir(iniPath), sourcePaths)
}

func assertTouchProfileBundleAllowed(root string, sourcePaths []string) error {
	detection, err := inspectTouchProfileBundle(root, sourcePaths)
	if err != nil {
		return err
	}
	if detection.Status == "none" {
		return nil
	}
	absolute, _ := filepath.Abs(root)
	code, message := "TOUCH_PROFILE_INPUT_ALREADY_TOUCH", "Input is already a Nahida Touch Profile mod and cannot be converted again."
	if detection.Status == "suspected" {
		code, message = "TOUCH_PROFILE_INPUT_SUSPECTED_TOUCH", "Input appears to be an existing Touch mod and cannot be safely converted."
	}
	details := ""
	if len(detection.Reasons) > 0 {
		details = " " + strings.Join(detection.Reasons, "; ") + "."
	}
	return contractError(fmt.Sprintf("%s: %s%s Path: %s", code, message, details, absolute))
}

func inspectTouchProfileBundle(root string, sourcePaths []string) (touchInputDetection, error) {
	root, _ = filepath.Abs(root)
	if raw, err := os.ReadFile(filepath.Join(root, touchProfileManifestFile)); err == nil {
		var manifest struct {
			Kind string `json:"kind"`
		}
		if json.Unmarshal(raw, &manifest) == nil && manifest.Kind == touchProfileManifestKind {
			return touchDetectionWithFiles(root, touchInputDetection{Status: "generated", Reasons: []string{"Nahida Touch Profile manifest found"}}), nil
		}
	}
	var builder strings.Builder
	for _, path := range sourcePaths {
		raw, err := os.ReadFile(path)
		if err != nil {
			return touchInputDetection{}, err
		}
		builder.Write(raw)
		builder.WriteByte('\n')
	}
	text := builder.String()
	states := touchMarkerNamespaces(text, regexp.MustCompile(`(?i)Nahida Touch Profile state\s*\(\s*([^\s)]+)\s*\)`))
	runtimes := touchMarkerNamespaces(text, regexp.MustCompile(`(?i)Nahida Touch Profile runtime\s*\(\s*([^\s)]+)\s*\)`))
	for _, state := range states {
		for _, runtime := range runtimes {
			if state == runtime {
				return touchDetectionWithFiles(root, touchInputDetection{Status: "generated", Reasons: []string{"Nahida Touch Profile INI markers found"}, Namespace: state}), nil
			}
		}
	}
	if len(states) > 0 || len(runtimes) > 0 {
		return touchDetectionWithFiles(root, touchInputDetection{Status: "incomplete", Reasons: []string{"Incomplete Nahida Touch Profile INI markers found"}}), nil
	}
	missing := missingTouchShaders(root)
	lower := strings.ToLower(text)
	references := true
	for _, name := range []string{"rzm_gs_probe.hlsl", "rzm_object_detect.hlsl", "rzm_jiggle_interaction.hlsl"} {
		references = references && strings.Contains(lower, name)
	}
	if len(missing) == 0 && references {
		return touchInputDetection{Status: "suspected", Reasons: []string{"Touch runtime shaders and INI references found"}}, nil
	}
	return touchInputDetection{Status: "none", Reasons: []string{}}, nil
}

func touchDetectionWithFiles(root string, input touchInputDetection) touchInputDetection {
	missing := missingTouchShaders(root)
	if len(missing) > 0 {
		input.Status = "incomplete"
		input.Reasons = append(input.Reasons, "Missing runtime shaders: "+strings.Join(missing, ", "))
	}
	return input
}
func missingTouchShaders(root string) []string {
	missing := []string{}
	for _, name := range touchShaderFiles {
		if info, err := os.Stat(filepath.Join(root, "Resources", "IM", name)); err != nil || !info.Mode().IsRegular() {
			missing = append(missing, name)
		}
	}
	return missing
}
func touchMarkerNamespaces(text string, re *regexp.Regexp) []string {
	matches := re.FindAllStringSubmatch(text, -1)
	out := make([]string, 0, len(matches))
	for _, match := range matches {
		out = append(out, strings.ToLower(match[1]))
	}
	return out
}
